<a href="https://colab.research.google.com/github/Isuru401/Statistical-Learning-e22401/blob/main/Gaussian_Mixture_Clustering_Conditional_Updating.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Gaussian Mixture Clustering as Conditional Updating

This notebook provides the complete mathematical derivations for Parts 1--9 and a computational implementation for Part 10.

We consider observations

$$
x_1,x_2,\ldots,x_n\in\mathbb{R}^d
$$

and latent cluster labels

$$
C_i\in\{1,\ldots,K\}.
$$

The prior cluster probabilities and conditional Gaussian model are

$$
P(C_i=k)=\phi_k,
\qquad
X_i\mid C_i=k\sim\mathcal{N}(\mu_k,\Sigma_k),
$$

where

$$
\phi_k\geq 0,
\qquad
\sum_{k=1}^{K}\phi_k=1.
$$

> **Important correction:** Since \(X_i\) is continuous, Bayes' rule must use probability densities such as \(p(x_i\mid C_i=k)\), rather than the point probability \(P(X_i=x_i)\).

## Part 1 — Deriving the Marginal Density

The latent variable can take one of the values \(1,\ldots,K\). Therefore, using the law of total probability for densities,

$$
p(x_i)
=
\sum_{k=1}^{K}p(x_i,C_i=k).
$$

Using the product rule,

$$
p(x_i,C_i=k)
=
p(x_i\mid C_i=k)P(C_i=k).
$$

Hence,

$$
p(x_i)
=
\sum_{k=1}^{K}
p(x_i\mid C_i=k)P(C_i=k).
$$

The model assumptions give

$$
P(C_i=k)=\phi_k
$$

and

$$
p(x_i\mid C_i=k)
=
\mathcal{N}(x_i\mid\mu_k,\Sigma_k).
$$

Therefore,

$$
\boxed{
p(x_i)
=
\sum_{k=1}^{K}
\phi_k\,
\mathcal{N}(x_i\mid\mu_k,\Sigma_k)
}.
$$

The multivariate Gaussian density is

$$
\mathcal{N}(x_i\mid\mu_k,\Sigma_k)
=
\frac{1}
{(2\pi)^{d/2}|\Sigma_k|^{1/2}}
\exp\left[
-\frac{1}{2}
(x_i-\mu_k)^T
\Sigma_k^{-1}
(x_i-\mu_k)
\right].
$$

This is called a **Gaussian mixture density** because it is a weighted combination of \(K\) Gaussian component densities. The coefficients \(\phi_k\) are nonnegative and add to one. Although every component is Gaussian, the complete mixture does not need to be a single Gaussian. It can be multimodal, asymmetric, and can contain components with different centers, spreads, and orientations.

## Part 2 — Deriving the Posterior Cluster Probability

For a fixed observation \(x_i\), Bayes' rule for a continuous observation gives

$$
P(C_i=k\mid X_i=x_i)
=
\frac{
p(x_i\mid C_i=k)P(C_i=k)
}{
p(x_i)
}.
$$

From Part 1,

$$
p(x_i)
=
\sum_{j=1}^{K}
p(x_i\mid C_i=j)P(C_i=j).
$$

Thus,

$$
P(C_i=k\mid X_i=x_i)
=
\frac{
p(x_i\mid C_i=k)P(C_i=k)
}{
\sum_{j=1}^{K}
p(x_i\mid C_i=j)P(C_i=j)
}.
$$

Substituting the Gaussian density and the prior cluster weight,

$$
\boxed{
P(C_i=k\mid X_i=x_i)
=
\frac{
\phi_k\mathcal{N}(x_i\mid\mu_k,\Sigma_k)
}{
\sum_{j=1}^{K}
\phi_j\mathcal{N}(x_i\mid\mu_j,\Sigma_j)
}
}.
$$

Define

$$
\boxed{
\gamma_{ik}
=
P(C_i=k\mid X_i=x_i)
}.
$$

Then

$$
\boxed{
\gamma_{ik}
=
\frac{
\phi_k\mathcal{N}(x_i\mid\mu_k,\Sigma_k)
}{
\sum_{j=1}^{K}
\phi_j\mathcal{N}(x_i\mid\mu_j,\Sigma_j)
}
}.
$$

The responsibilities satisfy

$$
0\leq\gamma_{ik}\leq 1
$$

and

$$
\sum_{k=1}^{K}\gamma_{ik}=1.
$$

The quantity \(\gamma_{ik}\) is a posterior probability because it updates the prior belief \(\phi_k\) after observing how likely \(x_i\) is under the Gaussian distribution of cluster \(k\).

## Part 3 — One-Hot Encoding of the Latent Cluster Variable

Define

$$
Z_i
=
\begin{bmatrix}
Z_{i1}\\
Z_{i2}\\
\vdots\\
Z_{iK}
\end{bmatrix},
$$

where

$$
Z_{ik}
=
\begin{cases}
1, & C_i=k,\\
0, & C_i\neq k.
\end{cases}
$$

Since \(Z_{ik}\) is an indicator random variable,

$$
\begin{aligned}
\mathbb{E}[Z_{ik}\mid X_i=x_i]
&=
1\cdot P(Z_{ik}=1\mid X_i=x_i)\\
&\quad+
0\cdot P(Z_{ik}=0\mid X_i=x_i).
\end{aligned}
$$

Therefore,

$$
\mathbb{E}[Z_{ik}\mid X_i=x_i]
=
P(Z_{ik}=1\mid X_i=x_i).
$$

The event \(Z_{ik}=1\) is the same as the event \(C_i=k\). Consequently,

$$
\mathbb{E}[Z_{ik}\mid X_i=x_i]
=
P(C_i=k\mid X_i=x_i)
=
\gamma_{ik}.
$$

Hence,

$$
\boxed{
\mathbb{E}[Z_{ik}\mid X_i=x_i]
=
\gamma_{ik}
}.
$$

Applying the result component-by-component,

$$
\begin{aligned}
\mathbb{E}[Z_i\mid X_i=x_i]
&=
\begin{bmatrix}
\mathbb{E}[Z_{i1}\mid X_i=x_i]\\
\mathbb{E}[Z_{i2}\mid X_i=x_i]\\
\vdots\\
\mathbb{E}[Z_{iK}\mid X_i=x_i]
\end{bmatrix}\\[4pt]
&=
\begin{bmatrix}
\gamma_{i1}\\
\gamma_{i2}\\
\vdots\\
\gamma_{iK}
\end{bmatrix}.
\end{aligned}
$$

Thus,

$$
\boxed{
\mathbb{E}[Z_i\mid X_i=x_i]
=
\begin{bmatrix}
\gamma_{i1}\\
\gamma_{i2}\\
\vdots\\
\gamma_{iK}
\end{bmatrix}
}.
$$

Therefore, the soft assignment in a GMM is precisely the conditional expectation of the one-hot latent membership vector.

## Part 4 — From Soft Assignment to Hard Clustering

The vector

$$
\mathbb{E}[Z_i\mid X_i=x_i]
=
(\gamma_{i1},\ldots,\gamma_{iK})^T
$$

retains the probability assigned to every cluster. This is a **soft assignment**.

A hard cluster label is obtained by selecting the largest posterior probability:

$$
\boxed{
\widehat C_i
=
\operatorname*{arg\,max}_{1\leq k\leq K}
\gamma_{ik}
}.
$$

For example, suppose

$$
(\gamma_{i1},\gamma_{i2},\gamma_{i3})
=
(0.55,0.40,0.05).
$$

The soft result shows meaningful uncertainty between clusters 1 and 2. The hard assignment is simply

$$
\widehat C_i=1.
$$

Therefore, soft clustering preserves uncertainty, while hard clustering converts the probability vector into one final label.

## Part 5 — Conditional Expectation of the Observation Given the Cluster

The model assumes

$$
X_i\mid C_i=k
\sim
\mathcal{N}(\mu_k,\Sigma_k).
$$

The expectation of a multivariate Gaussian random vector is its mean vector. Hence,

$$
\boxed{
\mathbb{E}[X_i\mid C_i=k]
=
\mu_k
}.
$$

The vector \(\mu_k\) is the average location of observations generated by cluster \(k\), so it is interpreted as the center of that cluster.

The two conditional expectations answer different questions:

$$
\mathbb{E}[Z_i\mid X_i=x_i]
=
\begin{bmatrix}
\gamma_{i1}\\
\vdots\\
\gamma_{iK}
\end{bmatrix}
$$

asks:

> Given the observed location \(x_i\), what are the posterior probabilities of the possible cluster labels?

It is a \(K\)-dimensional probability vector.

By contrast,

$$
\mathbb{E}[X_i\mid C_i=k]
=
\mu_k
$$

asks:

> Given cluster \(k\), where is an observation expected to be located?

It is a \(d\)-dimensional feature vector.

Therefore,

$$
\boxed{
\mathbb{E}[Z_i\mid X_i=x_i]
\text{ gives soft cluster membership}
}
$$

while

$$
\boxed{
\mathbb{E}[X_i\mid C_i=k]
\text{ gives the mean location of cluster }k.
}
$$

## Part 6 — The Complete-Data Likelihood

If both the observations and one-hot labels are known, the joint density for observation \(i\) is

$$
p(x_i,z_i)
=
\prod_{k=1}^{K}
\left[
\phi_k
\mathcal{N}(x_i\mid\mu_k,\Sigma_k)
\right]^{z_{ik}}.
$$

Only one \(z_{ik}\) equals one, so only the corresponding component contributes.

Assuming independent observations,

$$
\boxed{
p(x_1,\ldots,x_n,z_1,\ldots,z_n)
=
\prod_{i=1}^{n}
\prod_{k=1}^{K}
\left[
\phi_k
\mathcal{N}(x_i\mid\mu_k,\Sigma_k)
\right]^{z_{ik}}
}.
$$

Taking logarithms,

$$
\begin{aligned}
\ell_c
&=
\log
\left\{
\prod_{i=1}^{n}
\prod_{k=1}^{K}
\left[
\phi_k
\mathcal{N}(x_i\mid\mu_k,\Sigma_k)
\right]^{z_{ik}}
\right\}\\
&=
\sum_{i=1}^{n}
\sum_{k=1}^{K}
z_{ik}
\log
\left[
\phi_k
\mathcal{N}(x_i\mid\mu_k,\Sigma_k)
\right]\\
&=
\sum_{i=1}^{n}
\sum_{k=1}^{K}
z_{ik}
\left[
\log\phi_k
+
\log\mathcal{N}(x_i\mid\mu_k,\Sigma_k)
\right].
\end{aligned}
$$

Therefore,

$$
\boxed{
\ell_c
=
\sum_{i=1}^{n}
\sum_{k=1}^{K}
z_{ik}
\left[
\log\phi_k
+
\log\mathcal{N}(x_i\mid\mu_k,\Sigma_k)
\right]
}.
$$

If the values of \(z_{ik}\) were known, the observations belonging to each component would be known. The maximization would separate into simple cluster-specific calculations: proportions, means, and covariance matrices.

## Part 7 — The EM Interpretation

Let

$$
\theta
=
\{\phi_k,\mu_k,\Sigma_k\}_{k=1}^{K}
$$

denote all model parameters.

At iteration \(t\), the E-step computes

$$
\gamma_{ik}^{(t)}
=
P(C_i=k\mid X_i=x_i,\theta^{(t)}).
$$

Therefore,

$$
\boxed{
\gamma_{ik}^{(t)}
=
\frac{
\phi_k^{(t)}
\mathcal{N}
\left(
x_i\mid
\mu_k^{(t)},
\Sigma_k^{(t)}
\right)
}{
\sum_{j=1}^{K}
\phi_j^{(t)}
\mathcal{N}
\left(
x_i\mid
\mu_j^{(t)},
\Sigma_j^{(t)}
\right)
}
}.
$$

Because

$$
\mathbb{E}
[
Z_{ik}
\mid
X_i=x_i,\theta^{(t)}
]
=
\gamma_{ik}^{(t)},
$$

the unknown indicator is replaced by its conditional expectation:

$$
z_{ik}
\quad\leadsto\quad
\gamma_{ik}^{(t)}.
$$

The expected complete-data log-likelihood is

$$
Q(\theta\mid\theta^{(t)})
=
\mathbb{E}_{Z\mid X,\theta^{(t)}}
\left[
\ell_c(\theta)
\right].
$$

Therefore,

$$
\boxed{
Q(\theta\mid\theta^{(t)})
=
\sum_{i=1}^{n}
\sum_{k=1}^{K}
\gamma_{ik}^{(t)}
\left[
\log\phi_k
+
\log\mathcal{N}(x_i\mid\mu_k,\Sigma_k)
\right]
}.
$$

The E-step is a conditional update because it uses the current mixture weights and Gaussian parameters as the prior model, then updates cluster membership probabilities after each observation is considered.

## Part 8 — Parameter Updates

Define the effective membership of cluster \(k\) as

$$
\boxed{
N_k
=
\sum_{i=1}^{n}\gamma_{ik}
}.
$$

This value may be non-integer because observations can contribute fractionally.

### 8.1 Mixture-weight update

The terms involving \(\phi_k\) are

$$
Q_{\phi}
=
\sum_{k=1}^{K}N_k\log\phi_k.
$$

We maximize this subject to

$$
\sum_{k=1}^{K}\phi_k=1.
$$

Define the Lagrangian

$$
\mathcal{L}
=
\sum_{k=1}^{K}N_k\log\phi_k
+
\lambda
\left(
\sum_{k=1}^{K}\phi_k-1
\right).
$$

Differentiate:

$$
\frac{\partial\mathcal{L}}{\partial\phi_k}
=
\frac{N_k}{\phi_k}+\lambda.
$$

Setting the derivative to zero gives

$$
\phi_k=-\frac{N_k}{\lambda}.
$$

Also,

$$
\sum_{k=1}^{K}N_k
=
\sum_{i=1}^{n}\sum_{k=1}^{K}\gamma_{ik}
=
n.
$$

Using \(\sum_k\phi_k=1\),

$$
1=-\frac{n}{\lambda},
\qquad
\lambda=-n.
$$

Hence,

$$
\boxed{
\phi_k^{\mathrm{new}}
=
\frac{N_k}{n}
}.
$$

### 8.2 Mean update

Ignoring terms independent of \(\mu_k\),

$$
Q_{\mu_k}
=
-\frac{1}{2}
\sum_{i=1}^{n}
\gamma_{ik}
(x_i-\mu_k)^T
\Sigma_k^{-1}
(x_i-\mu_k).
$$

Differentiate:

$$
\frac{\partial Q_{\mu_k}}{\partial\mu_k}
=
\sum_{i=1}^{n}
\gamma_{ik}
\Sigma_k^{-1}
(x_i-\mu_k).
$$

Set this equal to zero:

$$
\sum_{i=1}^{n}
\gamma_{ik}
\Sigma_k^{-1}
(x_i-\mu_k)=0.
$$

Multiplying by \(\Sigma_k\),

$$
\sum_{i=1}^{n}\gamma_{ik}(x_i-\mu_k)=0.
$$

Therefore,

$$
\sum_{i=1}^{n}\gamma_{ik}x_i
=
\mu_k
\sum_{i=1}^{n}\gamma_{ik}
=
\mu_kN_k.
$$

Thus,

$$
\boxed{
\mu_k^{\mathrm{new}}
=
\frac{1}{N_k}
\sum_{i=1}^{n}
\gamma_{ik}x_i
}.
$$

### 8.3 Covariance update

Define

$$
S_k
=
\sum_{i=1}^{n}
\gamma_{ik}
(x_i-\mu_k^{\mathrm{new}})
(x_i-\mu_k^{\mathrm{new}})^T.
$$

The covariance-dependent terms can be written as

$$
Q_{\Sigma_k}
=
-\frac{1}{2}
\left[
N_k\log|\Sigma_k|
+
\operatorname{tr}
\left(
\Sigma_k^{-1}S_k
\right)
\right].
$$

Setting the matrix derivative to zero gives

$$
N_k\Sigma_k^{-1}
-
\Sigma_k^{-1}S_k\Sigma_k^{-1}
=
0.
$$

Multiplying appropriately by \(\Sigma_k\) gives

$$
N_k\Sigma_k=S_k.
$$

Therefore,

$$
\boxed{
\Sigma_k^{\mathrm{new}}
=
\frac{1}{N_k}
\sum_{i=1}^{n}
\gamma_{ik}
(x_i-\mu_k^{\mathrm{new}})
(x_i-\mu_k^{\mathrm{new}})^T
}.
$$

The responsibility \(\gamma_{ik}\) is a fractional membership weight:

- \(\gamma_{ik}\approx1\): strong contribution to cluster \(k\);
- \(\gamma_{ik}\approx0\): almost no contribution;
- \(0<\gamma_{ik}<1\): partial contribution.

Thus, the M-step uses posterior membership probabilities to calculate weighted cluster counts, means, and covariance matrices.

## Part 9 — Interpretation

Gaussian mixture clustering is a repeated process of conditional updating. The mixture weight \(\phi_k\) is the prior probability of cluster \(k\). The density

$$
\mathcal{N}(x_i\mid\mu_k,\Sigma_k)
$$

measures how compatible the observation \(x_i\) is with that cluster. Bayes' rule combines the prior weight and Gaussian compatibility to produce

$$
\gamma_{ik}
=
P(C_i=k\mid X_i=x_i),
$$

the posterior cluster probability after observing \(x_i\). The complete soft assignment is

$$
\mathbb{E}[Z_i\mid X_i=x_i]
=
(\gamma_{i1},\ldots,\gamma_{iK})^T.
$$

The M-step updates the cluster parameters by treating these posterior probabilities as fractional membership weights. These updated parameters are then used to calculate new posterior probabilities in the next E-step. Therefore, GMM clustering is probabilistic clustering based on repeated conditional expectations of hidden cluster-membership variables.

# Part 10 — Computational Simulation and Out-of-Sample Validation

The code uses two continuous variables from the credit-card dataset:

- `PURCHASES`
- `CREDIT_LIMIT`

The workflow is:

1. Split the observations into 80% training and 20% test data.
2. Learn missing-value imputation from the training set only.
3. Learn standardization from the training set only.
4. Fit a \(K=3\), full-covariance GMM using EM.
5. Calculate responsibilities with `predict_proba`.
6. Evaluate unseen observations using average test log-likelihood.
7. Generate three interactive Plotly figures.

In [1]:
# Install packages in Google Colab.
!pip -q install kagglehub plotly scikit-learn pandas numpy

In [2]:
# Download and load the dataset.
# A manual-upload fallback is included.

from pathlib import Path
import pandas as pd

try:
    import kagglehub

    dataset_directory = Path(
        kagglehub.dataset_download("arjunbhasin2013/ccdata")
    )

    csv_files = list(dataset_directory.rglob("*.csv"))
    if not csv_files:
        raise FileNotFoundError(
            f"No CSV file was found inside {dataset_directory}"
        )

    preferred_files = [
        path
        for path in csv_files
        if path.name.lower().replace("_", " ") == "cc general.csv"
    ]
    csv_path = preferred_files[0] if preferred_files else csv_files[0]

except Exception as error:
    print("Automatic Kaggle download was unavailable.")
    print("Reason:", error)
    print("Upload CC GENERAL.csv manually.")

    from google.colab import files

    uploaded = files.upload()
    if not uploaded:
        raise RuntimeError("No file was uploaded.")

    csv_path = Path(next(iter(uploaded.keys())))

credit_card_data = pd.read_csv(csv_path)

print("Loaded file:", csv_path)
print("Dataset shape:", credit_card_data.shape)
display(credit_card_data.head())

print("\nMissing values in the selected features:")
display(
    credit_card_data[["PURCHASES", "CREDIT_LIMIT"]]
    .isna()
    .sum()
    .to_frame("missing_count")
)

100%|██████████| 340k/340k [00:00<00:00, 21.7MB/s]

Extracting files...


Loaded file: /root/.cache/kagglehub/datasets/arjunbhasin2013/ccdata/versions/1/CC GENERAL.csv
Dataset shape: (8950, 18)


,CUST_ID,BALANCE,BALANCE_FREQUENCY,PURCHASES,ONEOFF_PURCHASES,INSTALLMENTS_PURCHASES,CASH_ADVANCE,PURCHASES_FREQUENCY,ONEOFF_PURCHASES_FREQUENCY,PURCHASES_INSTALLMENTS_FREQUENCY,CASH_ADVANCE_FREQUENCY,CASH_ADVANCE_TRX,PURCHASES_TRX,CREDIT_LIMIT,PAYMENTS,MINIMUM_PAYMENTS,PRC_FULL_PAYMENT,TENURE
0,C10001,40.900749,0.818182,95.40,0.00,95.4,0.000000,0.166667,0.000000,0.083333,0.000000,0,2,1000.0,201.802084,139.509787,0.000000,12
1,C10002,3202.467416,0.909091,0.00,0.00,0.0,6442.945483,0.000000,0.000000,0.000000,0.250000,4,0,7000.0,4103.032597,1072.340217,0.222222,12
2,C10003,2495.148862,1.000000,773.17,773.17,0.0,0.000000,1.000000,1.000000,0.000000,0.000000,0,12,7500.0,622.066742,627.284787,0.000000,12
3,C10004,1666.670542,0.636364,1499.00,1499.00,0.0,205.788017,0.083333,0.083333,0.000000,0.083333,1,1,7500.0,0.000000,NaN,0.000000,12
4,C10005,817.714335,1.000000,16.00,16.00,0.0,0.000000,0.083333,0.083333,0.000000,0.000000,0,1,1200.0,678.334763,244.791237,0.000000,12



Missing values in the selected features:


,missing_count
PURCHASES,0
CREDIT_LIMIT,1


In [3]:
import numpy as np
import pandas as pd

from sklearn.impute import SimpleImputer
from sklearn.mixture import GaussianMixture
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler

import plotly.express as px
import plotly.graph_objects as go


class GMMFinancialSegmenter:
    """
    Two-dimensional Gaussian Mixture Model for financial segmentation.

    The model uses training-only preprocessing, calculates posterior
    responsibilities, evaluates unseen data, and creates Plotly figures in
    the original feature units.
    """

    def __init__(
        self,
        feature_names=("PURCHASES", "CREDIT_LIMIT"),
        n_components=3,
        test_size=0.20,
        random_state=42,
        covariance_type="full",
        n_init=10,
        max_iter=500,
        reg_covar=1e-6,
        plot_quantiles=(0.01, 0.99),
    ):
        if len(feature_names) != 2:
            raise ValueError("Exactly two feature names must be supplied.")
        if n_components < 2:
            raise ValueError("n_components must be at least 2.")
        if not 0 < test_size < 1:
            raise ValueError("test_size must lie strictly between 0 and 1.")

        self.feature_names = tuple(feature_names)
        self.n_components = int(n_components)
        self.test_size = float(test_size)
        self.random_state = int(random_state)
        self.covariance_type = covariance_type
        self.n_init = int(n_init)
        self.max_iter = int(max_iter)
        self.reg_covar = float(reg_covar)
        self.plot_quantiles = plot_quantiles
        self._is_fitted = False

    def fit(self, data):
        """Fit preprocessing and the GMM using training data only."""
        missing_columns = [
            column
            for column in self.feature_names
            if column not in data.columns
        ]
        if missing_columns:
            raise KeyError(
                f"Dataset does not contain required columns: {missing_columns}"
            )

        selected_data = (
            data.loc[:, self.feature_names]
            .apply(pd.to_numeric, errors="coerce")
            .copy()
        )

        if selected_data.dropna(how="all").empty:
            raise ValueError("The selected features contain no usable data.")

        train_raw, test_raw = train_test_split(
            selected_data,
            test_size=self.test_size,
            random_state=self.random_state,
            shuffle=True,
        )

        self.imputer_ = SimpleImputer(strategy="median")
        train_imputed = self.imputer_.fit_transform(train_raw)
        test_imputed = self.imputer_.transform(test_raw)

        self.scaler_ = StandardScaler()
        train_scaled = self.scaler_.fit_transform(train_imputed)
        test_scaled = self.scaler_.transform(test_imputed)

        self.gmm_ = GaussianMixture(
            n_components=self.n_components,
            covariance_type=self.covariance_type,
            n_init=self.n_init,
            max_iter=self.max_iter,
            reg_covar=self.reg_covar,
            random_state=self.random_state,
            init_params="kmeans",
        )
        self.gmm_.fit(train_scaled)

        self.X_train_raw_ = pd.DataFrame(
            train_imputed,
            columns=self.feature_names,
            index=train_raw.index,
        )
        self.X_test_raw_ = pd.DataFrame(
            test_imputed,
            columns=self.feature_names,
            index=test_raw.index,
        )

        self.X_train_scaled_ = train_scaled
        self.X_test_scaled_ = test_scaled

        self.train_responsibilities_ = self.gmm_.predict_proba(train_scaled)
        self.test_responsibilities_ = self.gmm_.predict_proba(test_scaled)

        self.train_labels_ = self.train_responsibilities_.argmax(axis=1)
        self.test_labels_ = self.test_responsibilities_.argmax(axis=1)

        self.train_max_responsibility_ = (
            self.train_responsibilities_.max(axis=1)
        )
        self.test_max_responsibility_ = (
            self.test_responsibilities_.max(axis=1)
        )

        self.train_average_log_likelihood_ = self.gmm_.score(train_scaled)
        self.test_average_log_likelihood_ = self.gmm_.score(test_scaled)

        self._is_fitted = True

        print("=" * 62)
        print("GMM TRAINING REPORT")
        print("=" * 62)
        print(f"Converged: {self.gmm_.converged_}")
        print(f"EM iterations: {self.gmm_.n_iter_}")
        print(
            "Average training log-likelihood: "
            f"{self.train_average_log_likelihood_:.6f}"
        )
        print(
            "Average test log-likelihood: "
            f"{self.test_average_log_likelihood_:.6f}"
        )
        print(
            "Mean training maximum responsibility: "
            f"{self.train_max_responsibility_.mean():.6f}"
        )
        print(
            "Mean test maximum responsibility: "
            f"{self.test_max_responsibility_.mean():.6f}"
        )

        return self

    def _check_fitted(self):
        if not self._is_fitted:
            raise RuntimeError("Call fit(data) before using this method.")

    def transform(self, data):
        """Apply learned preprocessing to new observations."""
        self._check_fitted()

        selected = (
            data.loc[:, self.feature_names]
            .apply(pd.to_numeric, errors="coerce")
        )
        imputed = self.imputer_.transform(selected)
        return self.scaler_.transform(imputed)

    def predict_proba(self, data):
        """Return E[Z | X=x], the posterior responsibility vector."""
        transformed = self.transform(data)
        return self.gmm_.predict_proba(transformed)

    def predict(self, data):
        """Return argmax_k gamma_k(x)."""
        return self.predict_proba(data).argmax(axis=1)

    def cluster_summary(self):
        """Return mixture weights and centers in original units."""
        self._check_fitted()

        centers_raw = self.scaler_.inverse_transform(self.gmm_.means_)

        summary = pd.DataFrame(
            {
                "cluster": np.arange(self.n_components),
                "mixture_weight_phi": self.gmm_.weights_,
                f"center_{self.feature_names[0]}": centers_raw[:, 0],
                f"center_{self.feature_names[1]}": centers_raw[:, 1],
                "hard_training_count": np.bincount(
                    self.train_labels_,
                    minlength=self.n_components,
                ),
                "effective_training_membership_Nk": (
                    self.train_responsibilities_.sum(axis=0)
                ),
            }
        )
        return summary.sort_values("cluster").reset_index(drop=True)

    def responsibility_table(self, split="test", number_of_rows=10):
        """Display posterior probabilities for selected observations."""
        self._check_fitted()

        if split == "train":
            raw = self.X_train_raw_.copy()
            probabilities = self.train_responsibilities_
            labels = self.train_labels_
        elif split == "test":
            raw = self.X_test_raw_.copy()
            probabilities = self.test_responsibilities_
            labels = self.test_labels_
        else:
            raise ValueError("split must be either 'train' or 'test'.")

        table = raw.reset_index(drop=True)

        for component in range(self.n_components):
            table[f"gamma_{component}"] = probabilities[:, component]

        table["hard_cluster"] = labels
        table["max_responsibility"] = probabilities.max(axis=1)
        return table.head(number_of_rows)

    def plot_density_heatmap(self, nbins=60):
        """Empirical raw training-data density with marginal histograms."""
        self._check_fitted()

        x_feature, y_feature = self.feature_names

        figure = px.density_heatmap(
            self.X_train_raw_,
            x=x_feature,
            y=y_feature,
            nbinsx=nbins,
            nbinsy=nbins,
            histfunc="count",
            marginal_x="histogram",
            marginal_y="histogram",
            title=(
                "Empirical 2D Density Heatmap of Raw Training Data "
                "with Marginal Distributions"
            ),
            labels={"count": "Training observation count"},
        )

        figure.update_layout(
            template="plotly_white",
            height=750,
            title_x=0.5,
        )
        return figure

    def _create_grid(self, grid_size=180):
        """Calculate responsibilities on a grid in original units."""
        self._check_fitted()

        if grid_size < 50:
            raise ValueError("grid_size should be at least 50.")

        x_feature, y_feature = self.feature_names
        training = self.X_train_raw_

        if self.plot_quantiles is None:
            x_min = training[x_feature].min()
            x_max = training[x_feature].max()
            y_min = training[y_feature].min()
            y_max = training[y_feature].max()
        else:
            lower, upper = self.plot_quantiles
            if not 0 <= lower < upper <= 1:
                raise ValueError(
                    "plot_quantiles must satisfy 0 <= lower < upper <= 1."
                )

            x_min, x_max = training[x_feature].quantile([lower, upper])
            y_min, y_max = training[y_feature].quantile([lower, upper])

        x_margin = 0.05 * max(x_max - x_min, 1.0)
        y_margin = 0.05 * max(y_max - y_min, 1.0)

        x_values = np.linspace(
            max(0.0, x_min - x_margin),
            x_max + x_margin,
            grid_size,
        )
        y_values = np.linspace(
            max(0.0, y_min - y_margin),
            y_max + y_margin,
            grid_size,
        )

        grid_x, grid_y = np.meshgrid(x_values, y_values)
        grid_raw = np.column_stack([grid_x.ravel(), grid_y.ravel()])
        grid_scaled = self.scaler_.transform(grid_raw)

        responsibilities = self.gmm_.predict_proba(grid_scaled)

        maximum_responsibility = (
            responsibilities.max(axis=1).reshape(grid_x.shape)
        )
        hard_grid_label = (
            responsibilities.argmax(axis=1).reshape(grid_x.shape)
        )

        responsibility_text = np.array(
            [
                "<br>".join(
                    f"gamma_{component} = {value:.4f}"
                    for component, value in enumerate(probability_vector)
                )
                for probability_vector in responsibilities
            ]
        ).reshape(grid_x.shape)

        return {
            "x_values": x_values,
            "y_values": y_values,
            "maximum_responsibility": maximum_responsibility,
            "hard_grid_label": hard_grid_label,
            "responsibility_text": responsibility_text,
        }

    def _assignment_figure(self, split="train", grid_size=180):
        self._check_fitted()

        grid = self._create_grid(grid_size=grid_size)
        x_feature, y_feature = self.feature_names

        if split == "train":
            observations = self.X_train_raw_.copy()
            probabilities = self.train_responsibilities_
            labels = self.train_labels_
            title = (
                "Training Assignments over Maximum Posterior "
                "Responsibility Contours"
            )
        elif split == "test":
            observations = self.X_test_raw_.copy()
            probabilities = self.test_responsibilities_
            labels = self.test_labels_
            title = (
                "Out-of-Sample Test Assignments over Maximum Posterior "
                "Responsibility Contours"
            )
        else:
            raise ValueError("split must be either 'train' or 'test'.")

        figure = go.Figure()

        figure.add_trace(
            go.Contour(
                x=grid["x_values"],
                y=grid["y_values"],
                z=grid["maximum_responsibility"],
                text=grid["responsibility_text"],
                contours=dict(
                    start=1.0 / self.n_components,
                    end=1.0,
                    size=0.05,
                    coloring="heatmap",
                    showlines=False,
                ),
                colorscale="Viridis",
                opacity=0.72,
                colorbar=dict(title="max gamma", tickformat=".2f"),
                hovertemplate=(
                    f"{x_feature}: %{{x:.2f}}"
                    f"<br>{y_feature}: %{{y:.2f}}"
                    "<br>%{text}"
                    "<br>max gamma = %{z:.4f}"
                    "<extra></extra>"
                ),
                name="Maximum responsibility",
            )
        )

        if self.n_components > 1:
            figure.add_trace(
                go.Contour(
                    x=grid["x_values"],
                    y=grid["y_values"],
                    z=grid["hard_grid_label"],
                    contours=dict(
                        start=0.5,
                        end=self.n_components - 1.5,
                        size=1.0,
                        coloring="none",
                    ),
                    line=dict(width=2),
                    showscale=False,
                    hoverinfo="skip",
                    name="Hard decision boundaries",
                )
            )

        qualitative_colors = px.colors.qualitative.Safe

        for component in range(self.n_components):
            mask = labels == component
            component_observations = observations.loc[mask]
            component_probabilities = probabilities[mask]

            custom_data = np.column_stack(
                [
                    labels[mask],
                    component_probabilities.max(axis=1),
                ]
            )

            figure.add_trace(
                go.Scatter(
                    x=component_observations[x_feature],
                    y=component_observations[y_feature],
                    mode="markers",
                    name=f"Cluster {component}",
                    marker=dict(
                        size=6,
                        opacity=0.68,
                        color=qualitative_colors[
                            component % len(qualitative_colors)
                        ],
                        line=dict(width=0.3),
                    ),
                    customdata=custom_data,
                    hovertemplate=(
                        f"{x_feature}: %{{x:.2f}}"
                        f"<br>{y_feature}: %{{y:.2f}}"
                        "<br>Hard cluster: %{customdata[0]:.0f}"
                        "<br>max gamma: %{customdata[1]:.4f}"
                        "<extra></extra>"
                    ),
                )
            )

        figure.update_layout(
            title=title,
            title_x=0.5,
            template="plotly_white",
            height=760,
            xaxis_title=x_feature,
            yaxis_title=y_feature,
            legend_title="Hard assignment",
        )

        return figure

    def plot_training_assignments(self, grid_size=180):
        return self._assignment_figure(
            split="train",
            grid_size=grid_size,
        )

    def plot_test_assignments(self, grid_size=180):
        return self._assignment_figure(
            split="test",
            grid_size=grid_size,
        )

In [4]:
# Fit the requested K=3 GMM.

segmenter = GMMFinancialSegmenter(
    feature_names=("PURCHASES", "CREDIT_LIMIT"),
    n_components=3,
    test_size=0.20,
    random_state=42,
    covariance_type="full",
    n_init=10,
    max_iter=500,
    reg_covar=1e-6,
    # This affects only the displayed contour axes.
    # All observations are retained for fitting and evaluation.
    plot_quantiles=(0.01, 0.99),
).fit(credit_card_data)

GMM TRAINING REPORT
Converged: True
EM iterations: 19
Average training log-likelihood: -1.591637
Average test log-likelihood: -1.594085
Mean training maximum responsibility: 0.919921
Mean test maximum responsibility: 0.923041


In [5]:
# Learned prior weights and centers in original financial units.
display(segmenter.cluster_summary())

,cluster,mixture_weight_phi,center_PURCHASES,center_CREDIT_LIMIT,hard_training_count,effective_training_membership_Nk
0,0,0.444880,188.881562,2092.528699,3283,3161.246097
1,1,0.454698,928.916543,5801.914982,3251,3266.119500
2,2,0.100422,4839.529642,9571.026588,626,732.634403


In [6]:
# Posterior responsibilities for unseen observations.
# Every row is E[Z_i | X_i=x_i].
display(segmenter.responsibility_table(split="test", number_of_rows=12))

,PURCHASES,CREDIT_LIMIT,gamma_0,gamma_1,gamma_2,hard_cluster,max_responsibility
0,15.00,7500.0,4.897848e-05,0.974464,0.025487,1,0.974464
1,612.23,2000.0,7.922963e-01,0.204816,0.002887,0,0.792296
2,342.74,2000.0,9.681270e-01,0.031298,0.000575,0,0.968127
3,1901.71,1500.0,1.291413e-13,0.981648,0.018352,1,0.981648
4,3059.10,7000.0,2.930400e-46,0.486735,0.513265,2,0.513265
5,145.85,1000.0,9.810194e-01,0.018437,0.000543,0,0.981019
6,1160.09,2000.0,4.738329e-04,0.988022,0.011504,1,0.988022
7,885.44,1000.0,1.054016e-01,0.882523,0.012075,1,0.882523
8,760.38,4500.0,1.320162e-02,0.975185,0.011613,1,0.975185
9,0.00,8000.0,4.285880e-06,0.971953,0.028042,1,0.971953


In [7]:
# Figure 1: empirical density heatmap.
density_figure = segmenter.plot_density_heatmap(nbins=60)
density_figure.show()

In [8]:
# Figure 2: training assignments.
training_figure = segmenter.plot_training_assignments(grid_size=180)
training_figure.show()

In [ ]:
# Figure 3: out-of-sample test assignments.
test_figure = segmenter.plot_test_assignments(grid_size=180)
test_figure.show()

## Evaluation of the Computational Results

### 1. Empirical density heatmap

The density heatmap shows where the raw training observations are concentrated in the two-dimensional space formed by `PURCHASES` and `CREDIT_LIMIT`. Dense bins represent common financial behavior, while sparse regions represent less common or extreme customers. The marginal histograms show the separate distributions of each variable.

### 2. Training assignment plot

For every grid location \(x_{\mathrm{grid}}\), the model computes

$$
\mathbb{E}[Z\mid X=x_{\mathrm{grid}}]
=
\begin{bmatrix}
\gamma_1(x_{\mathrm{grid}})\\
\gamma_2(x_{\mathrm{grid}})\\
\gamma_3(x_{\mathrm{grid}})
\end{bmatrix}.
$$

The background displays

$$
\max_{1\leq k\leq3}
\gamma_k(x_{\mathrm{grid}}).
$$

A value close to one means a single component strongly dominates. A smaller value means that multiple components assign similar posterior probabilities to that location. Therefore, low-confidence bands around the white boundaries show regions of cluster ambiguity. Hovering over the contour displays the complete responsibility vector.

The point colors show the hard labels

$$
\widehat C_i
=
\operatorname*{arg\,max}_{k}\gamma_{ik}.
$$

Thus, the contour is a direct visual representation of the conditional expectation proved in Part 3.

### 3. Test assignment plot

The test observations were not used to learn the imputation values, scaling parameters, or GMM parameters. If test observations occupy similar high-confidence regions to the training observations, the learned mixture structure generalizes reasonably well. Test observations close to low-confidence boundaries naturally have ambiguous memberships.

### 4. Average test log-likelihood

The reported test score is

$$
\frac{1}{n_{\mathrm{test}}}
\sum_{i\in\mathrm{test}}
\log p(x_i).
$$

A larger value indicates that the fitted density assigns more probability density to unseen observations. Its absolute magnitude depends on preprocessing, features, covariance structure, and the number of components, so it is most useful for comparing models under the same setup.

### Reference result for these exact settings

With the specified 80/20 split, `random_state=42`, full covariance matrices, and three components, a reference execution converged in 19 EM iterations. The average test log-likelihood was approximately

$$
-1.5941.
$$

The component centers in the original variables were approximately

$$
(188.9,\ 2092.5),
\qquad
(928.9,\ 5801.9),
\qquad
(4839.5,\ 9571.0),
$$

where every ordered pair represents

$$
(\text{PURCHASES},\text{CREDIT\_LIMIT}).
$$

Component numbers are arbitrary. Another execution may permute the component labels without changing the underlying fitted mixture.

## Final Conclusion

The computational experiment confirms the theoretical conditional-updating interpretation. For each observation or grid location, the GMM calculates

$$
\mathbb{E}[Z\mid X=x],
$$

the posterior responsibility vector. The M-step then uses these responsibilities as fractional weights to update the component probabilities, means, and covariance matrices. Repeating the E-step and M-step creates a probabilistic clustering method that preserves uncertainty rather than immediately forcing every observation into one deterministic cluster.